### PR TITLE
Add rule deprecation mechanism and deprecate DuplicateCaseInWhenExpression, MissingWhenCase, RedundantElseInWhen

### PR DESCRIPTION
**Pull request by VitalyVPinchuk**
Close (#4973)
~~Exactly the same as `deprecation.properties` save for filename `deprecation.rules`~~.
Uses `deprecation.properties` to write deprecated rules.

### CODE REVIEW

File: detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/config/validation/ConfigValidation.kt [hunk 1]
```diff
@@ -84,14 +84,16 @@ private fun validateYamlConfig(
     baseline: YamlConfig,
     excludePatterns: Set<Regex>
 ): List<Notification> {
-    val deprecatedProperties = loadDeprecations()
+    val deprecatedProperties = loadDeprecations("deprecation.properties")
+    val deprecatedRules = loadDeprecations("deprecation.rules")
```
Comment (cortinico) [inline detekt-core/src/main/kotlin/io/gitlab/arturbosch/detekt/core/config/validation/ConfigValidation.kt:88]: There is no need of a separate file. We can keep on using `deprecation.properties` for this use case as well

File: detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/printer/DeprecatedPrinter.kt [hunk 2]
```diff
@@ -25,3 +25,23 @@ private fun writeProperty(ruleSet: RuleSetPage, rule: Rule, configuration: Confi
     @Suppress("UnsafeCallOnNullableType")
     return "${ruleSet.ruleSet.name}>${rule.name}>${configuration.name}=${configuration.deprecated!!}"
 }
+
+object DeprecatedRulePrinter : DocumentationPrinter<List<RuleSetPage>> {
+    @Suppress("NestedBlockDepth")
+    override fun print(item: List<RuleSetPage>): String {
+        val builder = StringBuilder()
+        item.forEach { ruleSet ->
+            ruleSet.rules.forEach { rule ->
```
Comment (schalkms) [inline detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/printer/DeprecatedPrinter.kt:34]: You could use `.filter()` first, then iterate over the rule. This way, you could spare the `@Suppress("NestedBlockDepth")` annotation.

**Review by schalkms — COMMENTED** [review]
Thanks for that! Looks good!
I also prefer using the existing single file for this use case, as Nico mentioned.

**Comment by VitalyVPinchuk** [comment]
Updated, and now it uses single file.

File: detekt-core/src/main/resources/deprecation.properties [hunk 3]
```diff
@@ -1,7 +1,11 @@
 complexity>LongParameterList>threshold=Use `functionThreshold` and `constructorThreshold` instead
 empty-blocks>EmptyFunctionBlock>ignoreOverriddenFunctions=Use `ignoreOverridden` instead
+potential-bugs>DuplicateCaseInWhenExpression=Compiler performs this check by default
```
Comment (cortinico) [inline detekt-core/src/main/resources/deprecation.properties:3]: Maybe reword to: 
```suggestion
potential-bugs>DuplicateCaseInWhenExpression=Rule deprecated as Compiler performs this check by default
```

File: detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/Rule.kt [hunk 4]
```diff
@@ -13,5 +13,8 @@ data class Rule(
     val configuration: List<Configuration> = emptyList(),
     val autoCorrect: Boolean = false,
     var inMultiRule: String? = null,
-    val requiresTypeResolution: Boolean = false
-)
+    val requiresTypeResolution: Boolean = false,
+    val deprecated: String? = null
```
Comment (cortinico) [inline detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/Rule.kt:17]:
```suggestion
    val deprecationMessage: String? = null
```

File: detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/DuplicateCaseInWhenExpression.kt [hunk 5]
```diff
@@ -32,6 +32,7 @@ import org.jetbrains.kotlin.psi.KtWhenExpression
  * </compliant>
  */
 @ActiveByDefault(since = "1.0.0")
+@Deprecated("Compiler performs this check by default")
```
Comment (cortinico) [inline detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/DuplicateCaseInWhenExpression.kt:35]: Nit: I think we should update the website to mention a note if a rule is deprecated.

File: detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/Rule.kt (hunk 4 above)
Comment (VitalyVPinchuk) [reply detekt-generator/src/main/kotlin/io/gitlab/arturbosch/detekt/generator/collection/Rule.kt]: I made it similar to `Configuration.deprecated`. 
Should I also reword that one?

File: detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/DuplicateCaseInWhenExpression.kt (hunk 5 above)
Comment (VitalyVPinchuk) [reply detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/DuplicateCaseInWhenExpression.kt]: Sure!
Updated `potential-bugs.md` in `versionized_docs`.

**Comment by cortinico** [comment]
You'll have to suppress the deprecation locally inside `PotentialBugProvider` as otherwise the CI is red.

```
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PotentialBugProvider.kt: (23, 13): 'DuplicateCaseInWhenExpression' is deprecated. Rule deprecated as compiler performs this check by default
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PotentialBugProvider.kt: (24, 13): 'ElseCaseInsteadOfExhaustiveWhen' is deprecated. Rule deprecated as compiler performs this check by default
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PotentialBugProvider.kt: (37, 13): 'MissingWhenCase' is deprecated. Rule deprecated as compiler performs this check by default
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PotentialBugProvider.kt: (39, 13): 'RedundantElseInWhen' is deprecated. Rule deprecated as compiler performs this check by default
```

**Comment by VitalyVPinchuk** [comment]
> You'll have to suppress the deprecation locally inside `PotentialBugProvider` as otherwise the CI is red.
> 
> ```
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PotentialBugProvider.kt: (23, 13): 'DuplicateCaseInWhenExpression' is deprecated. Rule deprecated as compiler performs this check by default
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PotentialBugProvider.kt: (24, 13): 'ElseCaseInsteadOfExhaustiveWhen' is deprecated. Rule deprecated as compiler performs this check by default
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PotentialBugProvider.kt: (37, 13): 'MissingWhenCase' is deprecated. Rule deprecated as compiler performs this check by default
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/PotentialBugProvider.kt: (39, 13): 'RedundantElseInWhen' is deprecated. Rule deprecated as compiler performs this check by default
> ```

I did it like this, but it doesn't help 
```kotlin
    @Suppress("DEPRECATION")
    override fun instance(config: Config): RuleSet = RuleSet(
```

**Comment by cortinico** [comment]
I think you'll have to inline it:

```kotlin
        listOf(
            AvoidReferentialEquality(config),
            Deprecation(config),
            DontDowncastCollectionTypes(config),
            DoubleMutabilityForCollection(config),
            @Suppress("DEPRECATION") DuplicateCaseInWhenExpression(config),
            ElseCaseInsteadOfExhaustiveWhen(config),
            EqualsAlwaysReturnsTrueOrFalse(config),
```

**Comment by cortinico** [comment]
You have more warnings:
```
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/DuplicateCaseInWhenExpressionSpec.kt: (9, 27): 'DuplicateCaseInWhenExpression' is deprecated. Rule deprecated as compiler performs this check by default
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ElseCaseInsteadOfExhaustiveWhenSpec.kt: (13, 27): 'ElseCaseInsteadOfExhaustiveWhen' is deprecated. Rule deprecated as compiler performs this check by default
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt: (17, 31): 'MissingWhenCase' is deprecated. Rule deprecated as compiler performs this check by default
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt: (312, 31): 'MissingWhenCase' is deprecated. Rule deprecated as compiler performs this check by default
w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/RedundantElseInWhenSpec.kt: (13, 27): 'RedundantElseInWhen' is deprecated. Rule deprecated as compiler performs this check by default
```

**Comment by VitalyVPinchuk** [comment]
> I think you'll have to inline it:
> 
> ```kotlin
>         listOf(
>             AvoidReferentialEquality(config),
>             Deprecation(config),
>             DontDowncastCollectionTypes(config),
>             DoubleMutabilityForCollection(config),
>             @Suppress("DEPRECATION") DuplicateCaseInWhenExpression(config),
>             ElseCaseInsteadOfExhaustiveWhen(config),
>             EqualsAlwaysReturnsTrueOrFalse(config),
> ```

Yes, that worked. Thanks!
This way suppressed rules are ignored by `RuleSetProviderCollector.kt`, so I slightly modified `visitCallExpression`.
```kotlin
?.map { if (it is KtAnnotatedExpression) it.lastChild as KtCallExpression else it }
```

**Comment by VitalyVPinchuk** [comment]
> You have more warnings:
> 
> ```
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/DuplicateCaseInWhenExpressionSpec.kt: (9, 27): 'DuplicateCaseInWhenExpression' is deprecated. Rule deprecated as compiler performs this check by default
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/ElseCaseInsteadOfExhaustiveWhenSpec.kt: (13, 27): 'ElseCaseInsteadOfExhaustiveWhen' is deprecated. Rule deprecated as compiler performs this check by default
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt: (17, 31): 'MissingWhenCase' is deprecated. Rule deprecated as compiler performs this check by default
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt: (312, 31): 'MissingWhenCase' is deprecated. Rule deprecated as compiler performs this check by default
> w: /home/runner/work/detekt/detekt/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/RedundantElseInWhenSpec.kt: (13, 27): 'RedundantElseInWhen' is deprecated. Rule deprecated as compiler performs this check by default
> ```

Right.
I didn't see them while running `./gradlew clean build`. Maybe there's a simple command to purge Gradle's cache without downloading dependencies?

**Comment by cortinico** [comment]
> I didn't see them while running `./gradlew clean build`. Maybe there's a simple command to purge Gradle's cache without downloading dependencies?

You should be able to replicate locally with:

```
./gradlew build -x detekt -PwarningsAsErrors=true
```

File: detekt-core/src/main/resources/default-detekt-config.yml [hunk 6]
```diff
@@ -419,10 +419,6 @@ potential-bugs:
       - 'java.util.HashSet'
       - 'java.util.LinkedHashMap'
       - 'java.util.HashMap'
-  DuplicateCaseInWhenExpression:
-    active: true
-  ElseCaseInsteadOfExhaustiveWhen:
-    active: false
```
Comment (BraisGabin) [inline detekt-core/src/main/resources/default-detekt-config.yml:425]: Should we deprecate this one? I know, we talked in the issue about "deprecate all the when related rules". But I feel that this one still have sense. The compiler doesn't check this. Is more, the compiler hints you to do this if you just want to remove the compiler error.

File: website/versioned_docs/version-1.21.0/rules/potential-bugs.md [hunk 7]
```diff
@@ -156,6 +156,8 @@ var myMap = mapOf("answer" to 42)
 
 ### DuplicateCaseInWhenExpression
 
+**Deprecated**: The rule is deprecated as compiler performs this check by default.
+
```
Comment (BraisGabin) [inline website/versioned_docs/version-1.21.0/rules/potential-bugs.md:160]: 🤔 Should we mark this as deprecated? I mean, they were not deprecated at version 1.21.0

File: detekt-generator/src/test/resources/RuleSet.md [hunk 8]
```diff
@@ -75,3 +75,23 @@ fun stuff(): Unit {}
...
+
+```kotlin
+fun stuff(): Unit {}
+```
+
+#### Compliant Code:
+
+```kotlin
+fun stuff() {}
+```
```
Comment (BraisGabin) [inline detekt-generator/src/test/resources/RuleSet.md:97]: Probably I'm loosing something here, but shouldn't this case have a "Deprecated"

File: detekt-generator/src/test/resources/RuleSet.md [hunk 9]
```diff
@@ -75,3 +75,23 @@ fun stuff(): Unit {}
 ```kotlin
 fun stuff() {}
 ```
+
+### DuplicateCaseInWhenExpression
```
Comment (BraisGabin) [inline detekt-generator/src/test/resources/RuleSet.md:79]: What do you think about line-cross the name of the Rule to make it clear that it is Deprecated.

### ~DuplicateCaseInWhenExpression~

```markdown
### ~DuplicateCaseInWhenExpression~
```

**Review by BraisGabin — COMMENTED** [review]
I like your solution :)!

But I really think that we should not deprecate `ElseCaseInsteadOfExhaustiveWhen`.

File: detekt-generator/src/test/resources/RuleSet.md (hunk 9 above)
Comment (cortinico) [reply detekt-generator/src/test/resources/RuleSet.md:79]: Cool idea :)

File: website/versioned_docs/version-1.21.0/rules/potential-bugs.md (hunk 7 above)
Comment (cortinico) [reply website/versioned_docs/version-1.21.0/rules/potential-bugs.md:160]: Agree, let's not update the docs for 1.21.0

File: detekt-generator/src/test/resources/RuleSet.md (hunk 8 above)
Comment (VitalyVPinchuk) [reply detekt-generator/src/test/resources/RuleSet.md:97]: Hey @BraisGabin 
There is `DeprecatedPrinterSpec` test, which tests deprecated configurations. So I added a deprecated rule in the test set `createRuleSetPage()` to verify the `DeprecatedPrinter` output.

The same set `createRuleSetPage()` is used in `RuleSetPagePrinterSpec` test, so to keep test working I had to modify the test reference file `RuleSet.md`.

**Comment by cortinico** [comment]
> Add rule deprecation mechanism

Can we also update the title of this PR mentioning which rule we deprecated?

File: website/versioned_docs/version-1.21.0/rules/potential-bugs.md (hunk 7 above)
Comment (VitalyVPinchuk) [reply website/versioned_docs/version-1.21.0/rules/potential-bugs.md]: Reverted.

File: detekt-core/src/main/resources/default-detekt-config.yml (hunk 6 above)
Comment (VitalyVPinchuk) [reply detekt-core/src/main/resources/default-detekt-config.yml:425]: Reverted.

**Review by BraisGabin — APPROVED** [review]
LGTM Thank you so much for the contribution.
